tests/test_remote.py

```diff
@@ -0,0 +1,225 @@
+# Copyright 2018-2020 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Unit tests for the Fock plugin.
+"""
+import pytest
+
+import strawberryfields as sf
+from strawberryfields.api import Result
+import pennylane_sf
+
+import pennylane as qml
+import numpy as np
+
+
+class TestDevice:
+    """General tests for the HardwareDevice."""
+
+    def test_token(self, monkeypatch):
+        """Tests that the SF store_account function is called with token."""
+        test_token = "SomeToken"
+        recorder = []
+        monkeypatch.setattr("strawberryfields.store_account", lambda arg: recorder.append(arg))
+        dev = qml.device('strawberryfields.ai', backend="X8", wires=8, shots=10, sf_token=test_token)
+
+        assert recorder[0] == test_token
+
+
+MOCK_SAMPLES = np.array([[3, 4, 2, 3, 4, 3, 1, 0],
+       [4, 3, 3, 2, 0, 3, 1, 4],
+       [2, 1, 3, 3, 3, 2, 2, 4],
+       [4, 1, 4, 4, 2, 3, 3, 0],
+       [4, 2, 3, 3, 3, 0, 0, 4],
+       [1, 2, 4, 4, 2, 0, 0, 4],
+       [2, 3, 1, 2, 1, 0, 4, 1],
+       [1, 2, 0, 1, 2, 3, 3, 0],
+       [1, 2, 4, 0, 0, 4, 2, 4],
+       [1, 0, 1, 1, 1, 3, 1, 0]])
+
+MOCK_SAMPLES_PROD = np.array([0, 0, 864, 0, 0, 0, 0, 0, 0, 0])
+
+class MockEngine:
+    """Mock SF engine class"""
+
+    def __init__(*args):
+        pass
+
+    def run(*args, **kwargs):
+        return Result(MOCK_SAMPLES)
+
+class TestSample:
+    """Tests that samples are correctly returned from the hardware device."""
+
+    def test_mocked_engine_run_samples(self, monkeypatch):
+        """Tests that samples are determined by the RemoteEngine.run method
+        from SF by using a mocked RemoteEngine."""
+        modes = 8
+        shots = 10
+        dev = qml.device('strawberryfields.ai', backend="X8", wires=modes, shots=shots)
+
+        @qml.qnode(dev)
+        def quantum_function(theta, phi):
+            qml.Beamsplitter(theta, phi,wires=[0,1])
+            qml.Beamsplitter(theta, phi,wires=[4,5])
+            return qml.sample(qml.TensorN(wires=list(range(modes))))
```

Comment:
Thinking about this some more, would `return [qml.sample(qml.NumberOperator(i)) for i in range(wires)]` correspond to the fock basis samples?

Reply:
Yes. That's a good point, that this actually returns a single measurement outcome for each shot.

A circuit with `shots=10`
* `return qml.sample(qml.TensorN(wires=list(range(modes))))` could return

```python
array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0])
```
and means `10` circuit evaluations (for `10` shots).

Whereas with
* `return [qml.sample(qml.NumberOperator(i)) for i in range(8)]` could return

```python
array([[1, 0, 1, 0, 0, 0, 1, 1, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 4, 0, 0],
       [0, 1, 0, 0, 1, 0, 0, 1, 0, 0],
       [0, 0, 0, 1, 0, 0, 0, 0, 0, 0],
       [0, 0, 1, 0, 0, 0, 0, 0, 1, 0],
       [0, 0, 0, 1, 0, 0, 0, 0, 0, 1],
       [0, 0, 0, 0, 0, 0, 0, 2, 0, 2],
       [0, 0, 0, 0, 2, 0, 1, 1, 0, 1]])
```

In the latter case, we'd, however, have `8*10` circuit evaluations. With the current logic we "dispose" of the samples obtained for `wires!=i`. Perhaps it would be cool to store the samples? :thinking: